test/org/javarosa/xpath/expr/XPathPathExprCurrentItemsetNodesetTest.java

```diff
@@ -0,0 +1,84 @@
+ * limitations under the License.
+ */
+package org.javarosa.xpath.expr;
+
+import org.javarosa.core.model.FormDef;
+import org.javarosa.core.model.ItemsetBinding;
+import org.javarosa.core.model.SelectChoice;
+import org.javarosa.core.model.data.SelectOneData;
+import org.javarosa.core.model.data.StringData;
+import org.javarosa.core.model.data.helper.Selection;
+import org.javarosa.core.model.instance.InstanceInitializationFactory;
+import org.javarosa.core.test.FormParseInit;
+import org.javarosa.form.api.FormEntryController;
+import org.javarosa.form.api.FormEntryModel;
+import org.javarosa.form.api.FormEntryPrompt;
+import org.junit.Before;
+import org.junit.Test;
+
+import static junit.framework.TestCase.assertEquals;
+import static org.javarosa.test.utils.ResourcePathHelper.r;
+
+public class XPathPathExprCurrentItemsetNodesetTest {
+    private FormDef formDef;
+    private FormEntryController formEntryController;
+
+    @Before
+    public void setUp() {
+        FormParseInit fpi = new FormParseInit(r("relative-current-ref-itemset-nodeset.xml"));
+        formDef = fpi.getFormDef();
+        formDef.initialize(true, new InstanceInitializationFactory());
+        FormEntryModel formEntryModel = new FormEntryModel(formDef);
+        formEntryController = new FormEntryController(formEntryModel);
+    }
+
+    /**
+     * current() in an itemset nodeset expression should refer to the select node. This is verified by building an
+     * itemset from a repeat which is a sibling of the select.
+     */
+    @Test
+    public void current_in_itemset_nodeset_should_refer_to_node() {
+        // don't know how to jump to repeat directly so jump to following question and step backwards
+        formEntryController.jumpToFirstQuestionWithName("selected_person");
+        formEntryController.stepToPreviousEvent(); // repeat
+        formEntryController.descendIntoRepeat(0);
+        formEntryController.stepToNextEvent(); // person
+
+        StringData nameValue = new StringData("Bob");
+        formEntryController.answerQuestion(nameValue, true);
+
+        formEntryController.stepToNextEvent();
+        formEntryController.descendIntoNewRepeat();
+        formEntryController.stepToNextEvent(); // person
+
+        StringData nameValue2 = new StringData("Janet");
+        formEntryController.answerQuestion(nameValue2, true);
+
+        formEntryController.jumpToFirstQuestionWithName("selected_person");
+        FormEntryPrompt personPrompt = formEntryController.getModel().getQuestionPrompt();
+        personPrompt.getAnswerValue();
+        ItemsetBinding dynamicChoices = personPrompt.getQuestion().getDynamicChoices();
+
+        SelectChoice personChoice = dynamicChoices.getChoices().get(1);
+        assertEquals("Janet", personPrompt.getSelectChoiceText(personChoice));
+        SelectOneData personSelection = new SelectOneData(new Selection(personChoice));
+        formEntryController.answerQuestion(personSelection, true);
```

Comment:
~It's hard to figure out what we're asserting without studying the blank form. Also, why are we jumping to the same question we answered with "Bob" and "Janet" in the first part? Some comments would help.~

It took me several times going back and forth and read the form definition to understand what was going on. The most confusing part for me was to see the same "selected_person" on lines 55 and 70. I though we were answering "selected_person" from lines 55 to 70, but now I understand that we were using that question as our first pivot point in the form's DOM to navigate to the "/data/people" repeat. 

Also, is really necessary to actually make a selection here? Would it be enough to just check that there are two choices and that they match "Bob" and "Janet"?

Reply:
This is the same pattern as https://github.com/opendatakit/javarosa/pull/335#discussion_r201705617. I didn't feel totally confident in the test until I knew the round trip worked.